### ISSUE TITLE
Local reification doesn't respect `NoFieldSelectors`

### ISSUE THREAD

**Issue by RyanGlScott** [goldfirere/th-desugar#183]
In order to reproduce this bug, you'll need a recent version of GHC HEAD that includes [this commit](https://gitlab.haskell.org/ghc/ghc/-/commit/3f374399e2dbebcdfe5bc31f94fc502b46d0cf4f), which fixes a key interaction between `NoFieldSelectors` and Template Haskell. If you run this program:

```hs
{-# LANGUAGE NoFieldSelectors #-}
{-# LANGUAGE TemplateHaskell #-}
module Main where

import Language.Haskell.TH
import Language.Haskell.TH.Desugar

data T1 = MkT1 { unT1 :: Int }

$(pure [])

main :: IO ()
main = do
  putStrLn $(lookupValueName "unT1" >>= stringE . show)
  putStrLn $(do decs <- [d| data T2 = MkT2 { unT2 :: Int } |]
                withLocalDeclarations decs (lookupValueNameWithLocals "unT2") >>= stringE . show)
```

Then you will get the following output:

```
Nothing
Just unT2_6989586621679020976
```

This is surprising, as I would have expected both lines to be `Nothing`. Because `NoFieldSelectors` is enabled, neither `unT1` (which is defined outside of TH quotes) nor `unT2` (which is defined in a local TH quote) should be treated as a top-level field selector. Despite this, `lookupValueNameWithLocals` incorrectly finds `unT2` as if it were a top-level name.

**Comment by goldfirere** [comment]
I agree with your conclusion, but I want to note that there is wiggle-room here. Consider the possibility that a quote is created in a different file from where it is processed. Which set of extensions should apply? One could argue that the extensions in effect where the quote appears are the more relevant ones, because the quote author has control over those extensions. But Template Haskell in its current form has no affordances to allow this to work, so we're left using the extensions at process-time, which is what you propose. So, maybe there isn't actually any wiggle room, but I claim that the current design of this aspect of TH is squishy. Ideally, an untyped TH quote would carry some metadata saying what extensions were in force at the quote site, but that's not the TH we have.

**Comment by RyanGlScott** [comment]
Indeed, this is squishy. We are doing this sort thing already with `MonadFailDesugaring` (see [here](https://github.com/goldfirere/th-desugar/blob/dbae660a0f507122b0ea04b6f0c0d8141342c21e/Language/Haskell/TH/Desugar/Core.hs#L522-L524)), so we aren't really treading new ground here. Still, this squishiness would be worth mentioning in the limitations section of the `README`.

**Comment by RyanGlScott** [comment]
Fixed in #197.
